### PR TITLE
CDRIVER-3991 Add versioned API connection examples

### PR DESCRIPTION
**Pull request by alcaeus**
CDRIVER-3991

### CODE REVIEW

File: src/libmongoc/tests/test-mongoc-sample-commands.c [hunk 1]
```diff
@@ -3572,6 +3572,134 @@ callback (mongoc_client_session_t *session,
...
+
+   /* For a replica set, include the replica set name and a seedlist of the
+    * members in the URI string; e.g.
+    * uri_repl = "mongodb://mongodb0.example.com:27017,mongodb1.example.com:" \
+    *    "27017/?replicaSet=myRepl";
+    * client = test_framework_client_new (uri_repl);
+    * For a sharded cluster, connect to the mongos instances; e.g.
+    * uri_sharded =
+    * "mongodb://mongos0.example.com:27017,mongos1.example.com:27017/";
+    * client = test_framework_client_new (uri_sharded);
```
Comment (kevinAlbs) [inline src/libmongoc/tests/test-mongoc-sample-commands.c:3592]:
```suggestion
    * client = mongoc_client_new (uri_sharded);
```

File: src/libmongoc/tests/test-mongoc-sample-commands.c [hunk 2]
```diff
@@ -3572,6 +3572,134 @@ callback (mongoc_client_session_t *session,
...
+   mongoc_client_t *client = NULL;
+   mongoc_server_api_t *server_api = NULL;
+   mongoc_server_api_version_t server_api_version;
+   bson_error_t error;
+
+   /* For a replica set, include the replica set name and a seedlist of the
+    * members in the URI string; e.g.
+    * uri_repl = "mongodb://mongodb0.example.com:27017,mongodb1.example.com:" \
+    *    "27017/?replicaSet=myRepl";
+    * client = test_framework_client_new (uri_repl);
```
Comment (kevinAlbs) [inline src/libmongoc/tests/test-mongoc-sample-commands.c:3588]: Though I see this in other examples, it seems wrong to refer to a test function in examples meant for the public docs.

```suggestion
    * client = mongoc_client_new (uri_repl);
```

File: src/libmongoc/tests/test-mongoc-sample-commands.c [hunk 3]
```diff
@@ -3572,6 +3572,134 @@ callback (mongoc_client_session_t *session,
...
```
Comment (kevinAlbs) [inline src/libmongoc/tests/test-mongoc-sample-commands.c:3617]:
```suggestion
    * client = mongoc_client_new (uri_repl);
```

File: src/libmongoc/tests/test-mongoc-sample-commands.c [hunk 4]
```diff
@@ -3572,6 +3572,134 @@ callback (mongoc_client_session_t *session,
...
+    * client = test_framework_client_new (uri_sharded);
+    */
+
+   client = get_client ();
+
+   mongoc_server_api_version_from_string ("1", &server_api_version);
+   server_api = mongoc_server_api_new (server_api_version);
+
+   mongoc_client_set_server_api (client, server_api, &error);
+   /* End Versioned API Example 1 */
```
Comment (kevinAlbs) [inline src/libmongoc/tests/test-mongoc-sample-commands.c:3601]: Needs a `mongoc_client_destroy (client)` and `mongoc_server_api_destroy (server_api)`.

File: src/libmongoc/tests/test-mongoc-sample-commands.c [hunk 5]
```diff
@@ -3572,6 +3572,134 @@ callback (mongoc_client_session_t *session,
...
+    * client = test_framework_client_new (uri_sharded);
+    */
+
+   client = get_client ();
+
+   mongoc_server_api_version_from_string ("1", &server_api_version);
+   server_api = mongoc_server_api_new (server_api_version);
+   mongoc_server_api_strict (server_api, true);
+
+   mongoc_client_set_server_api (client, server_api, &error);
```
Comment (kevinAlbs) [inline src/libmongoc/tests/test-mongoc-sample-commands.c:3630]: Needs a `mongoc_client_destroy (client)` and `mongoc_server_api_destroy (server_api)`.

File: src/libmongoc/tests/test-mongoc-sample-commands.c [hunk 6]
```diff
@@ -3572,6 +3572,134 @@ callback (mongoc_client_session_t *session,
...
+
+   /* For a replica set, include the replica set name and a seedlist of the
+    * members in the URI string; e.g.
+    * uri_repl = "mongodb://mongodb0.example.com:27017,mongodb1.example.com:" \
+    *    "27017/?replicaSet=myRepl";
+    * client = test_framework_client_new (uri_repl);
+    * For a sharded cluster, connect to the mongos instances; e.g.
+    * uri_sharded =
+    * "mongodb://mongos0.example.com:27017,mongos1.example.com:27017/";
+    * client = test_framework_client_new (uri_sharded);
```
Comment (kevinAlbs) [inline src/libmongoc/tests/test-mongoc-sample-commands.c:3621]:
```suggestion
    * client = mongoc_client_new (uri_sharded);
```

File: src/libmongoc/tests/test-mongoc-sample-commands.c [hunk 7]
```diff
@@ -3572,6 +3572,134 @@ callback (mongoc_client_session_t *session,
...
+   mongoc_client_t *client = NULL;
+   mongoc_server_api_t *server_api = NULL;
+   mongoc_server_api_version_t server_api_version;
+   bson_error_t error;
+
+   /* For a replica set, include the replica set name and a seedlist of the
+    * members in the URI string; e.g.
+    * uri_repl = "mongodb://mongodb0.example.com:27017,mongodb1.example.com:" \
+    *    "27017/?replicaSet=myRepl";
+    * client = test_framework_client_new (uri_repl);
```
Comment (kevinAlbs) [inline src/libmongoc/tests/test-mongoc-sample-commands.c:3647]:
```suggestion
    * client = mongoc_client_new (uri_repl);
```

File: src/libmongoc/tests/test-mongoc-sample-commands.c [hunk 8]
```diff
@@ -3572,6 +3572,134 @@ callback (mongoc_client_session_t *session,
...
+    * For a sharded cluster, connect to the mongos instances; e.g.
+    * uri_sharded =
+    * "mongodb://mongos0.example.com:27017,mongos1.example.com:27017/";
+    * client = test_framework_client_new (uri_sharded);
```
Comment (kevinAlbs) [inline src/libmongoc/tests/test-mongoc-sample-commands.c:3651]:
```suggestion
    * client = mongoc_client_new (uri_sharded);
```

File: src/libmongoc/tests/test-mongoc-sample-commands.c [hunk 9]
```diff
@@ -3572,6 +3572,134 @@ callback (mongoc_client_session_t *session,
...
+    */
+
+   client = get_client ();
+
+   mongoc_server_api_version_from_string ("1", &server_api_version);
+   server_api = mongoc_server_api_new (server_api_version);
+   mongoc_server_api_strict (server_api, false);
+
+   mongoc_client_set_server_api (client, server_api, &error);
+   /* End Versioned API Example 3 */
```
Comment (kevinAlbs) [inline src/libmongoc/tests/test-mongoc-sample-commands.c:3661]: Needs a `mongoc_client_destroy (client)` and `mongoc_server_api_destroy (server_api)`.

File: src/libmongoc/tests/test-mongoc-sample-commands.c [hunk 10]
```diff
@@ -3572,6 +3572,134 @@ callback (mongoc_client_session_t *session,
...
+   mongoc_client_t *client = NULL;
+   mongoc_server_api_t *server_api = NULL;
+   mongoc_server_api_version_t server_api_version;
+   bson_error_t error;
+
+   /* For a replica set, include the replica set name and a seedlist of the
+    * members in the URI string; e.g.
+    * uri_repl = "mongodb://mongodb0.example.com:27017,mongodb1.example.com:" \
+    *    "27017/?replicaSet=myRepl";
+    * client = test_framework_client_new (uri_repl);
```
Comment (kevinAlbs) [inline src/libmongoc/tests/test-mongoc-sample-commands.c:3677]:
```suggestion
    * client = mongoc_client_new (uri_repl);
```

File: src/libmongoc/tests/test-mongoc-sample-commands.c [hunk 11]
```diff
@@ -3572,6 +3572,134 @@ callback (mongoc_client_session_t *session,
...
+    * For a sharded cluster, connect to the mongos instances; e.g.
+    * uri_sharded =
+    * "mongodb://mongos0.example.com:27017,mongos1.example.com:27017/";
+    * client = test_framework_client_new (uri_sharded);
```
Comment (kevinAlbs) [inline src/libmongoc/tests/test-mongoc-sample-commands.c:3681]:
```suggestion
    * client = mongoc_client_new (uri_sharded);
```

File: src/libmongoc/tests/test-mongoc-sample-commands.c [hunk 12]
```diff
@@ -3572,6 +3572,134 @@ callback (mongoc_client_session_t *session,
...
+    */
+
+   client = get_client ();
+
+   mongoc_server_api_version_from_string ("1", &server_api_version);
+   server_api = mongoc_server_api_new (server_api_version);
+   mongoc_server_api_deprecation_errors (server_api, true);
+
+   mongoc_client_set_server_api (client, server_api, &error);
+   /* End Versioned API Example 4 */
```
Comment (kevinAlbs) [inline src/libmongoc/tests/test-mongoc-sample-commands.c:3691]: Needs a `mongoc_client_destroy (client)` and `mongoc_server_api_destroy (server_api)`.

File: src/libmongoc/tests/test-mongoc-sample-commands.c [hunk 13]
```diff
@@ -3648,6 +3776,7 @@ test_sample_commands (void)
    if (!test_framework_max_wire_version_at_least (WIRE_VERSION_4_4)) {
       test_sample_txn_commands (client);
    }
+   test_sample_versioned_api ();
```
Comment (kevinAlbs) [inline src/libmongoc/tests/test-mongoc-sample-commands.c:3779]: This should skip on pre-5.0 servers.

File: src/libmongoc/tests/test-mongoc-sample-commands.c (hunk 4 above)
Comment (alcaeus) [reply src/libmongoc/tests/test-mongoc-sample-commands.c:3601]: Added in all 4 tests (outside of the sample code)

File: src/libmongoc/tests/test-mongoc-sample-commands.c (hunk 1 above)
Comment (alcaeus) [reply src/libmongoc/tests/test-mongoc-sample-commands.c]: Ah, great catch. I copied this from the transactions tests where this is also part of the public sample. I've changed that occurrence as well.

File: src/libmongoc/tests/test-mongoc-sample-commands.c (hunk 13 above)
Comment (alcaeus) [reply src/libmongoc/tests/test-mongoc-sample-commands.c]: Done

**Review by kevinAlbs — APPROVED** [review]
LGTM

**Pull request merged**